In [5]:
# checking installations
import numpy
print(numpy.__version__)


2.1.0


In [6]:
# importing packages
import os
import numpy as np
import pandas as pd
import scipy
import anndata
import scanpy as sc
import pybiomart
import scvi
import torch
import random
import seaborn as sns

In [7]:
# using getcwd() to retrive the current working directory where the script is running
# storing directory path as a string in a variable called cwd
cwd = os.getcwd()
cwd

'/Users/hiraali/Desktop/FigureOneLab/trevino'

# Storing Cell Metadata

Pulled cell metadata file through supplementary website: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE162170

In [15]:
# read a compressed metadata file located inside the cwd into a Pandas DataFrame called meta.
meta = pd.read_csv(cwd+'/data/GSE162170_rna_cell_metadata.txt.gz', compression='gzip', sep='\t')
# Set the DataFrame (meta) index to the values in the 'Cell.ID' column
# now we can reference rows using Cell.ID instead of default numerical indices.
meta.index = meta['Cell.ID']
# remove the index name (usually to make things cleaner)
meta.index.name = None
# printing meta
meta

Cell.ID      Sample.ID  \
hft_w20_p3_r1_AAACCCAAGCTGCGAA  hft_w20_p3_r1_AAACCCAAGCTGCGAA  hft_w20_p3_r1   
hft_w20_p3_r1_AAACCCAAGGTAGTAT  hft_w20_p3_r1_AAACCCAAGGTAGTAT  hft_w20_p3_r1   
hft_w20_p3_r1_AAACCCACAACTCCAA  hft_w20_p3_r1_AAACCCACAACTCCAA  hft_w20_p3_r1   
hft_w20_p3_r1_AAACCCACATAGTCAC  hft_w20_p3_r1_AAACCCACATAGTCAC  hft_w20_p3_r1   
hft_w20_p3_r1_AAACCCAGTACAGGTG  hft_w20_p3_r1_AAACCCAGTACAGGTG  hft_w20_p3_r1   
...                                                        ...            ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA  hft_w16_p7_r2_TTTGTTGCAGCACCCA  hft_w16_p7_r2   
hft_w16_p7_r2_TTTGTTGCAGGCTACC  hft_w16_p7_r2_TTTGTTGCAGGCTACC  hft_w16_p7_r2   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA  hft_w16_p7_r2_TTTGTTGGTCGCTTAA  hft_w16_p7_r2   
hft_w16_p7_r2_TTTGTTGGTCGTACAT  hft_w16_p7_r2_TTTGTTGGTCGTACAT  hft_w16_p7_r2   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG  hft_w16_p7_r2_TTTGTTGGTTAGTTCG  hft_w16_p7_r2   

                                  Age Tissue.ID Sample.Type   Assay     Batch  \
hft_w20_p3_r1_AAACCCAAGCTGCGAA  pcw20      HFT3         HFT  RNA v3  b2019_06   
hft_w20_p3_r1_AAACCCAAGGTAGTAT  pcw20      HFT3         HFT  RNA v3  b2019_06   
hft_w20_p3_r1_AAACCCACAACTCCAA  pcw20      HFT3         HFT  RNA v3  b2019_06   
hft_w20_p3_r1_AAACCCACATAGTCAC  pcw20      HFT3         HFT  RNA v3  b2019_06   
hft_w20_p3_r1_AAACCCAGTACAGGTG  pcw20      HFT3         HFT  RNA v3  b2019_06   
...                               ...       ...         ...     ...       ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA  pcw16      HFT7         HFT  RNA NG  b2020_03   
hft_w16_p7_r2_TTTGTTGCAGGCTACC  pcw16      HFT7         HFT  RNA NG  b2020_03   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA  pcw16      HFT7         HFT  RNA NG  b2020_03   
hft_w16_p7_r2_TTTGTTGGTCGTACAT  pcw16      HFT7         HFT  RNA NG  b2020_03   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG  pcw16      HFT7         HFT  RNA NG  b2020_03   

                               seurat_clusters  RNA.Counts  RNA.Features  ...  \
hft_w20_p3_r1_AAACCCAAGCTGCGAA             c16        1397           677  ...   
hft_w20_p3_r1_AAACCCAAGGTAGTAT             c11       14338          4301  ...   
hft_w20_p3_r1_AAACCCACAACTCCAA             c17        9260          3481  ...   
hft_w20_p3_r1_AAACCCACATAGTCAC              c0        4025          1969  ...   
hft_w20_p3_r1_AAACCCAGTACAGGTG              c4        7131          2930  ...   
...                                        ...         ...           ...  ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA              c6        8532          3497  ...   
hft_w16_p7_r2_TTTGTTGCAGGCTACC              c2        6689          2393  ...   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA              c2        3865          1799  ...   
hft_w16_p7_r2_TTTGTTGGTCGTACAT              c2        5293          2365  ...   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG              c6        5291          2530  ...   

                                    Cell.Barcode   DF_pANN DF_classification  \
hft_w20_p3_r1_AAACCCAAGCTGCGAA  AAACCCAAGCTGCGAA  0.086339           Singlet   
hft_w20_p3_r1_AAACCCAAGGTAGTAT  AAACCCAAGGTAGTAT  0.325683           Singlet   
hft_w20_p3_r1_AAACCCACAACTCCAA  AAACCCACAACTCCAA  0.397814           Doublet   
hft_w20_p3_r1_AAACCCACATAGTCAC  AAACCCACATAGTCAC  0.076503           Singlet   
hft_w20_p3_r1_AAACCCAGTACAGGTG  AAACCCAGTACAGGTG  0.239344           Singlet   
...                                          ...       ...               ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA  TTTGTTGCAGCACCCA  0.389488           Doublet   
hft_w16_p7_r2_TTTGTTGCAGGCTACC  TTTGTTGCAGGCTACC  0.210916           Singlet   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA  TTTGTTGGTCGCTTAA  0.074798           Singlet   
hft_w16_p7_r2_TTTGTTGGTCGTACAT  TTTGTTGGTCGTACAT  0.208895           Singlet   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG  TTTGTTGGTTAGTTCG  0.313342           Singlet   

                                DF_pANN_quantile Spliced.Counts  \
hft_w20_p3_r1_AAACCCAAGCTGCGAA          0.356997           1063   
hft_w20_p3_r1_AAACCCAAGGTAGTAT          0

## Exploring the Metadata

Below, we preview the first few rows of the dataframe, meta. We see the following metadata columns stored:
* Cell.ID = i.e. hft_w20_p3_r1_AAACCCAAGCTGCGAA
* Sample.ID = i.e. hft_w20_p3_r1
* Age = i.e. pcw20 (the post-conceptional week time-point)
* Tissue.ID = i.e. HFT3	
* Batch = b2019_06
* seurat_clusters = i.e. c16
* RNA Counts = i.e. 1397
* RNA Features = i.e. 677
* Cell barcode = i.e. AAACCCAAGCTGCGAA	
* etc.

We can also get an overview of the dataframe (using .info()).

In [22]:
# preview the first few rows of the df, meta
meta.head()
# get an overview of the df, meta
meta.info()

<class 'pandas.core.frame.DataFrame'>
Index: 57868 entries, hft_w20_p3_r1_AAACCCAAGCTGCGAA to hft_w16_p7_r2_TTTGTTGGTTAGTTCG
Data columns (total 22 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Cell.ID             57868 non-null  object 
 1   Sample.ID           57868 non-null  object 
 2   Age                 57868 non-null  object 
 3   Tissue.ID           57868 non-null  object 
 4   Sample.Type         57868 non-null  object 
 5   Assay               57868 non-null  object 
 6   Batch               57868 non-null  object 
 7   seurat_clusters     57868 non-null  object 
 8   RNA.Counts          57868 non-null  int64  
 9   RNA.Features        57868 non-null  int64  
 10  Percent.MT          57868 non-null  float64
 11  Percent.Ribo        57868 non-null  float64
 12  Cell.Barcode        57868 non-null  object 
 13  DF_pANN             57868 non-null  float64
 14  DF_classification   57868 non-null  object 
 15  DF_p

# Storing RNA Counts

Pulled cell metadata file through supplementary website: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE162170


In [25]:
%%time
counts = pd.read_csv(cwd+'/data/GSE162170_rna_counts.tsv.gz', compression='gzip', sep='\t')
counts = counts.transpose()
counts

CPU times: user 5min 30s, sys: 29.5 s, total: 6min
Wall time: 8min 18s


ENSG00000243485  ENSG00000237613  \
hft_w20_p3_r1_AAACCCAAGCTGCGAA                0                0   
hft_w20_p3_r1_AAACCCAAGGTAGTAT                0                0   
hft_w20_p3_r1_AAACCCACAACTCCAA                0                0   
hft_w20_p3_r1_AAACCCACATAGTCAC                0                0   
hft_w20_p3_r1_AAACCCAGTACAGGTG                0                0   
...                                         ...              ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA                0                0   
hft_w16_p7_r2_TTTGTTGCAGGCTACC                0                0   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA                0                0   
hft_w16_p7_r2_TTTGTTGGTCGTACAT                0                0   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG                0                0   

                                ENSG00000186092  ENSG00000238009  \
hft_w20_p3_r1_AAACCCAAGCTGCGAA                0                0   
hft_w20_p3_r1_AAACCCAAGGTAGTAT                0                0   
hft_w20_p3_r1_AAACCCACAACTCCAA                0                0   
hft_w20_p3_r1_AAACCCACATAGTCAC                0                0   
hft_w20_p3_r1_AAACCCAGTACAGGTG                0                0   
...                                         ...              ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA                0                0   
hft_w16_p7_r2_TTTGTTGCAGGCTACC                0                0   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA                0                0   
hft_w16_p7_r2_TTTGTTGGTCGTACAT                0                0   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG                0                0   

                                ENSG00000239945  ENSG00000239906  \
hft_w20_p3_r1_AAACCCAAGCTGCGAA                0                0   
hft_w20_p3_r1_AAACCCAAGGTAGTAT                0                0   
hft_w20_p3_r1_AAACCCACAACTCCAA                0                0   
hft_w20_p3_r1_AAACCCACATAGTCAC                0                0   
hft_w20_p3_r1_AAACCCAGTACAGGTG                0                0   
...                                         ...              ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA                0                0   
hft_w16_p7_r2_TTTGTTGCAGGCTACC                0                0   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA                0                0   
hft_w16_p7_r2_TTTGTTGGTCGTACAT                0                0   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG                0                0   

                                ENSG00000241599  ENSG00000236601  \
hft_w20_p3_r1_AAACCCAAGCTGCGAA                0                0   
hft_w20_p3_r1_AAACCCAAGGTAGTAT                0                0   
hft_w20_p3_r1_AAACCCACAACTCCAA                0                0   
hft_w20_p3_r1_AAACCCACATAGTCAC                0                0   
hft_w20_p3_r1_AAACCCAGTACAGGTG                0                0   
...                                         ...              ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA                0                0   
hft_w16_p7_r2_TTTGTTGCAGGCTACC                0                0   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA                0                0   
hft_w16_p7_r2_TTTGTTGGTCGTACAT                0                0   
hft_w16_p7_r2_TTTGTTGGTTAGTTCG                0                0   

                                ENSG00000284733  ENSG00000235146  ...  \
hft_w20_p3_r1_AAACCCAAGCTGCGAA                0                0  ...   
hft_w20_p3_r1_AAACCCAAGGTAGTAT                0                0  ...   
hft_w20_p3_r1_AAACCCACAACTCCAA                0                0  ...   
hft_w20_p3_r1_AAACCCACATAGTCAC                0                0  ...   
hft_w20_p3_r1_AAACCCAGTACAGGTG                0                0  ...   
...                                         ...              ...  ...   
hft_w16_p7_r2_TTTGTTGCAGCACCCA                0                0  ...   
hft_w16_p7_r2_TTTGTTGCAGGCTACC                0                0  ...   
hft_w16_p7_r2_TTTGTTGGTCGCTTAA                0                0  ...   
hft_w16_p7_r2_TTTGTTGGTCGTACAT    

In [ ]:
%%time
adata = anndata.AnnData(X=counts,
                        obs=meta,
                        var=counts.columns.to_frame())
adata

In [ ]:
%%time
a = sc.queries.biomart_annotations('hsapiens', ['ensembl_gene_id','hgnc_symbol'])
b = dict(zip(a['ensembl_gene_id'], a['hgnc_symbol']))
adata.var['hgnc_symbol'] = adata.var[0].map(b)
adata.var.index = adata.var['hgnc_symbol']
adata.var.index.name = None
adata.var.drop(columns=[0,'hgnc_symbol'], inplace=True)
adata.var

In [ ]:
adata.var_names_make_unique

In [ ]:
a = ~adata.var.index.isnull()
adata = adata[:,a].copy()

In [ ]:
adata

In [ ]:
%%time
sc.pp.filter_genes(adata, min_cells=10)
sc.pp.filter_cells(adata, min_genes=200)

In [ ]:
#sc.pp.subsample(adata, n_obs=40000, random_state=0, copy=False)

In [ ]:
%%time
adata.X = scipy.sparse.csr_matrix(adata.X.copy())
adata.layers['counts'] = scipy.sparse.csr_matrix(adata.X.copy())
#adata.layers['log2_counts'] = sc.pp.log1p(adata.layers['counts'].copy(), base=2)

In [ ]:
%%time
random.seed(17)
scvi.model.SCVI.setup_anndata(adata, layer='counts', batch_key='Sample.ID')
scvi_model = scvi.model.SCVI(adata, n_layers=2, n_latent=30, n_hidden=128, gene_likelihood='nb')
scvi_model.train()

In [ ]:
%%time
random.seed(17)
adata.obsm['X_scvi'] = scvi_model.get_latent_representation()
adata.layers['counts_scvi'] = scvi_model.get_normalized_expression(library_size=10000)
#adata.layers['log2_counts_scvi'] = sc.pp.log1p(adata.layers['counts_scvi'].copy(), base=2)

In [ ]:
%%time
sc.pp.neighbors(adata, use_rep='X_scvi', key_added='neighbors_scvi', n_neighbors=20)
sc.tl.leiden(adata, neighbors_key='neighbors_scvi', key_added='leiden_scvi', resolution=3)
sc.tl.umap(adata, neighbors_key='neighbors_scvi')
sc.pl.umap(adata, color=['leiden_scvi'], legend_loc='on data')

In [ ]:
%%time
sc.pp.neighbors(adata, use_rep='X_scvi', key_added='neighbors_scvi', n_neighbors=20)
sc.tl.leiden(adata, neighbors_key='neighbors_scvi', key_added='leiden_scvi', resolution=3)
sc.tl.umap(adata, neighbors_key='neighbors_scvi')
sc.pl.umap(adata, color=['leiden_scvi'], legend_loc='on data')

In [ ]:
%%time
adata.write(cwd+'/outs/231226_trevino_rna_scvi.h5ad')